worldedit-core/src/main/java/com/sk89q/worldedit/extent/clipboard/io/BuiltInClipboardFormat.java

```diff
@@ -19,39 +19,105 @@
 import java.io.BufferedOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.util.Locale;
-import java.util.Map;
 import java.util.Set;
 import java.util.zip.GZIPInputStream;
 import java.util.zip.GZIPOutputStream;
 
 /**
  * A collection of supported clipboard formats.
  */
+@SuppressWarnings("removal") //FAWE: suppress JNBT deprecations
 public enum BuiltInClipboardFormat implements ClipboardFormat {
 
     //FAWE start - register fast clipboard io
-    FAST("fast", "fawe", "sponge", "schem") {
+    FAST_V3("fast", "fawe", "schem") {
+        @Override
+        public ClipboardReader getReader(InputStream inputStream) throws IOException {
+            return new FastSchematicReaderV3(inputStream);
+        }
+
+        @Override
+        public ClipboardWriter getWriter(OutputStream outputStream) throws IOException {
+            OutputStream gzip;
+            if (outputStream instanceof ParallelGZIPOutputStream || outputStream instanceof GZIPOutputStream) {
+                gzip = outputStream;
+            } else {
+                outputStream = new BufferedOutputStream(outputStream);
+                gzip = new ParallelGZIPOutputStream(outputStream);
+            }
+            NBTOutputStream nbtStream = new NBTOutputStream(new BufferedOutputStream(gzip));
+            return new FastSchematicWriterV3(nbtStream);
+        }
+
+        @Override
+        public boolean isFormat(final InputStream inputStream) {
+            try (final DataInputStream stream = new DataInputStream(new FastBufferedInputStream(new GZIPInputStream(inputStream)));
+                 final NBTInputStream nbt = new NBTInputStream(stream)) {
+                if (stream.readByte() != NBTConstants.TYPE_COMPOUND) {
+                    return false;
+                }
+                stream.skipNBytes(2); // TAG name length ("" = 0), no need to read name as no bytes are written for root tag
+                if (stream.readByte() != NBTConstants.TYPE_COMPOUND) {
+                    return false;
+                }
+                stream.skipNBytes(2); // TAG name length ("Schematic" = 9)
+                stream.skipNBytes(9); // "Schematic"
+
+                // We can't guarantee the specific order of nbt data, so scan and skip, if required
+                do {
+                    byte type = stream.readByte();
+                    String name = stream.readUTF();
+                    if (type == NBTConstants.TYPE_END) {
+                        return false;
+                    }
+                    if (type == NBTConstants.TYPE_INT && name.equals("Version")) {
+                        return stream.readInt() == FastSchematicWriterV3.CURRENT_VERSION;
+                    }
+                    nbt.readTagPayloadLazy(type, 0);
+                } while (true);
```

Comment:
I really hate this about the sponge format, but I guess letting people put whatever file extension they want is important for some reason...? Really stupid aspect of the format and I wouldn't be opposed to checking file extension == .schem or .sponge and maybe having a magic something that FAWE writes into the schematic we can check? Given the header of the schematic will always be the same (if written by FAWE) perhaps we can just check magic N bytes matches the start of the file, whether that basically includes the `"": { "Schematic": { "Version": 3` or whatever

Reply:
I guess for these few bytes, yea - simple BufferedReader#mark + reset should work

Reply:
We're still having to read the whole file though, I'm suggesting having two checks that will return much quicker - limiting the file extensions we support (as is done by literally everything - wrong extensions, assume wrong file) and checking for some "magic" that FAWE writes, as we have control over that, otherwise, use this method

Reply:
Exactly, the Stream is wrapped into a Buffered stream either way (iirc). Even when checking for the magic bytes (the schematic part you posted) it'd have to reset potentially. When the sequence does not match, the stream might be somewhere in between a compound or tag name.

Reply:
Yeah that's true, I would still add a file extension check though

Reply:
Yeah, might be worth doing:
> The highly recommended file extension for files using this specification is .schem

So I guess we can enforce that

Reply:
Or maybe we don't so we don't have to deal with the occasional person that tries to load a v2 or a v3 schematic with a .schematic extension because some schematic download sites are stupid